sklearn/neighbors/nca.py

```diff
@@ -0,0 +1,596 @@
+                    pca.fit(X)
+                    transformation = pca.components_
+                elif init == 'lda':
+                    from ..discriminant_analysis import (
+                        LinearDiscriminantAnalysis)
+                    lda = LinearDiscriminantAnalysis(n_components=n_components)
+                    if self.verbose:
+                        print('Finding most discriminative components... ',
+                              end='')
+                        sys.stdout.flush()
+                    lda.fit(X, y)
+                    transformation = lda.scalings_.T[:n_components]
+                if self.verbose:
+                    print('done in {:5.2f}s'.format(time.time() - init_time))
+        return transformation
+
+    def _callback(self, transformation):
+        """Called after each iteration of the optimizer.
+
+        Parameters
+        ----------
+        transformation : array, shape(n_components, n_features)
+            The solution computed by the optimizer in this iteration.
+        """
+        if self.callback is not None:
+            self.callback(transformation, self.n_iter_)
+
+        self.n_iter_ += 1
+
+    def _loss_grad_lbfgs(self, transformation, X, mask, sign=1.0):
+        """Compute the loss and the loss gradient w.r.t. ``transformation``.
+
+        Parameters
+        ----------
+        transformation : array, shape (n_components, n_features)
+            The linear transformation on which to compute loss and evaluate
+            gradient
+
+        X : array, shape (n_samples, n_features)
+            The training samples.
+
+        mask : array, shape (n_samples, n_samples)
+            A mask where ``mask[i, j] == 1`` if ``X[i]`` and ``X[j]`` belong
+            to the same class, and ``0`` otherwise.
+
+        Returns
+        -------
+        loss : float
+            The loss computed for the given transformation.
+
+        gradient : array, shape (n_components * n_features,)
+            The new (flattened) gradient of the loss.
+        """
+
+        if self.n_iter_ == 0:
+            self.n_iter_ += 1
+            if self.verbose:
+                header_fields = ['Iteration', 'Objective Value', 'Time(s)']
+                header_fmt = '{:>10} {:>20} {:>10}'
+                header = header_fmt.format(*header_fields)
+                cls_name = self.__class__.__name__
+                print('[{}]'.format(cls_name))
+                print('[{}] {}\n[{}] {}'.format(cls_name, header,
+                                                cls_name, '-' * len(header)))
+
+        t_funcall = time.time()
+
+        transformation = transformation.reshape(-1, X.shape[1])
+        X_embedded = np.dot(X, transformation.T)  # (n_samples, n_components)
+
+        # Compute softmax distances
+        p_ij = pairwise_distances(X_embedded, squared=True)
+        np.fill_diagonal(p_ij, np.inf)
+        p_ij = np.exp(-p_ij - logsumexp(-p_ij, axis=1)[:, np.newaxis])
```

Comment:
We have a softmax function in sklearn.utils.extmath, but it uses division for normalisation rather than logarithmic arithmetic. But it subtracts from all values to avoid overflow. Can you reuse that function? Should we be using logsumexp there?

Reply:
Thanks, I didn't know about this function. I replaced the current implementation by it and tests still pass, and running on the example it seems to work the exact same manner. I'll run a quick benchmark though because if I remember correctly I used `logsumexp` because it worked a bit better in some particular cases (that I can't remember), but this was probably before the factorization of the gradient (which introduce a little bit of uncertainties)...